Sources/LeafKit/LeafData/LeafDataStorage.swift

```diff
@@ -3,46 +3,220 @@
-    ///
-    ///     true
-    ///
+internal indirect enum LeafDataStorage: Equatable, CustomStringConvertible {
+    // MARK: - Cases
+    
+    // Static values
     case bool(Bool)
-
-    /// A `String`.
-    ///
-    ///     "hello"
-    ///
     case string(String)
-
-    /// An `Int`.
-    ///
-    ///     42
-    ///
     case int(Int)
-
-    /// A `Double`.
-    ///
-    ///     3.14
-    ///
     case double(Double)
-
-    /// `Data` blob.
-    ///
-    ///     Data([0x72, 0x73])
-    ///
     case data(Data)
-
-    /// A nestable `[String: LeafData]` dictionary.
+    
+    // Collections (potentially holding lazy values)
     case dictionary([String: LeafData])
-
-    /// A nestable `[LeafData]` array.
     case array([LeafData])
 
-    /// A lazily-resolvable `LeafData`.
-    case lazy(() -> (LeafData))
+    // Wrapped `Optional<LeafDataStorage>`
+    case optional(_ wrapped: LeafDataStorage?, _ type: LeafData.NaturalType)
+    
+    // Lazy resolvable function
+    // Must specify return tuple giving (returnType, invariance)
+    case lazy(f: () -> (LeafData),
+              returns: LeafData.NaturalType,
+              invariant: Bool)
+    
+    // MARK: - LeafSymbol Conformance
+    
+    // MARK: Properties
+    internal var resolved: Bool { true }
+    internal var invariant: Bool {
+        switch self {
+            case .lazy(_, _, let invariant): return invariant
+            case .optional(let o, _): return o?.invariant ?? true
+            case .array(let a):
+                let stored = a.map { $0.storage }.filter { $0.isLazy }
+                return stored.allSatisfy { $0.invariant }
+            case .dictionary(let d):
+                let stored = d.values.map { $0.storage }.filter { $0.isLazy }
+                return stored.allSatisfy { $0.invariant }
+            default: return true
+        }
+    }
+    internal var symbols: Set<String> { .init() }
+    internal var isAtomic: Bool { true }
+    internal var isExpression: Bool { false }
+    internal var isAny: Bool { false }
+    internal var isConcrete: Bool { true }
+    /// Note: Will *always* return a value - can be force-unwrapped safely
+    internal var concreteType: LeafData.NaturalType? {
+        switch self {
+            // Concrete Types
+            case .array(_)           : return .array
+            case .bool(_)            : return .bool
+            case .data(_)            : return .data
+            case .dictionary(_)      : return .dictionary
+            case .double(_)          : return .double
+            case .int(_)             : return .int
+            case .string(_)          : return .string
+            // Internal Wrapped Types
+            case .lazy(_, let t, _),
+                 .optional(_, let t) : return t
+        }
+    }
+    
+    // MARK: Functions
+    
+    /// Will resolve anything but variant Lazy data (99% of everything), and unwrap optionals
+    internal func softResolve() -> LeafDataStorage {
```

Comment:
Strictly speaking not relevant for LeafDataStorage. This is nearly always just an identity return. We never actually create a wrapped optional at this point - all .optional cases are going to be .none until Leaf syntax supports handling nils more directly.